src/shims/foreign_items.rs

```diff
@@ -105,27 +103,37 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
     /// Emulates calling a foreign item, failing if the item is not supported.
     /// This function will handle `goto_block` if needed.
+    /// Returns Ok(None) if the foreign item was completely handled
+    /// by this function.
+    /// Returns Ok(body) if processing the foreign item
+    /// requires us to call the function 'body'
     fn emulate_foreign_item(
         &mut self,
-        def_id: DefId,
+        link_name: &str,
         args: &[OpTy<'tcx, Tag>],
         dest: Option<PlaceTy<'tcx, Tag>>,
         ret: Option<mir::BasicBlock>,
-    ) -> InterpResult<'tcx> {
+        _unwind: Option<mir::BasicBlock>
+    ) -> InterpResult<'tcx, Option<&'mir mir::Body<'tcx>>> {
         let this = self.eval_context_mut();
-        let attrs = this.tcx.get_attrs(def_id);
-        let link_name = match attr::first_attr_value_str_by_name(&attrs, sym::link_name) {
-            Some(name) => name.as_str(),
-            None => this.tcx.item_name(def_id).as_str(),
-        };
-        // Strip linker suffixes (seen on 32-bit macOS).
-        let link_name = link_name.trim_end_matches("$UNIX2003");
-        let tcx = &{ this.tcx.tcx };
+        let tcx = &{this.tcx.tcx};
 
         // First: functions that diverge.
         match link_name {
-            "__rust_start_panic" | "panic_impl" => {
-                throw_unsup_format!("the evaluated program panicked");
+            "__rust_start_panic" => {
+                return Ok(Some(this.handle_rust_start_panic()?))
```

Comment:
At this point it seems easier to inline that 3-line function. When doing so, please add a comment saying that while this looks like we are implementing `__rust_start_panic` using `__rust_start_panic`, what really happens is that we implement calls to the `__rust_start_panic` FFI import using the `__rust_start_panic` function *exported* from the panic runtime.